pkg/envoy/envoy.go

```diff
@@ -80,6 +94,74 @@ func StartEnvoy(debug bool, adminPort int, stateDir, logDir string) *Envoy {
 	return e
 }
 
+// isEOF returns true if the error message ends in "EOF". ReadMsgUnix returns extra info in the beginning.
+func isEOF(err error) bool {
+	strerr := err.Error()
+	errlen := len(strerr)
+	return errlen >= 3 && strerr[errlen-3:] == io.EOF.Error()
+}
+
+// StopEnvoy kills the Envoy process started with StartEnvoy. The gRPC API streams are terminated
+// first.
+func (e *Envoy) StartAccessLogging(accessLogPath string) {
+	// Create the access log listener
+	os.Remove(accessLogPath)
+	var err error
+	e.accessLogListener, err = net.ListenUnix("unixpacket", &net.UnixAddr{Name: accessLogPath, Net: "unixpacket"})
+	if err != nil {
+		log.WithError(err).Fatal("Envoy: Failed to listen at ", accessLogPath)
+	}
+	e.accessLogListener.SetUnlinkOnClose(true)
+
+	go func(uxlis *net.UnixListener) {
+		for {
+			// Each Envoy listener opens a new connection over the Unix domain socket.
+			// Multiple worker threads serving the listener share that same connection
+			uc, err := uxlis.AcceptUnix()
+			if err != nil {
+				log.WithError(err).Error("AcceptUnix failed")
+				continue
+			}
+			log.Info("Envoy: Access log connection opened")
+			go func(conn *net.UnixConn) {
+				buf := make([]byte, 4096)
+				for {
+					n, _, flags, _, err := conn.ReadMsgUnix(buf, nil)
+					if err != nil {
+						if !isEOF(err) {
+							log.WithError(err).Error("Envoy: Access log read error")
+						}
+						break
+					}
+					if flags&syscall.MSG_TRUNC != 0 {
+						log.Warning("Envoy: Truncated access log message discarded.")
+						continue
+					}
+
+					pblog := HttpLogEntry{}
+					err = proto.Unmarshal(buf[:n], &pblog)
+					if err != nil {
+						log.WithError(err).Warning("Envoy: Invalid accesslog.proto HttpLogEntry message.")
+						continue
+					}
+
+					// Correlate the log entry with a listener
+					l := e.lds.findListener(pblog.CiliumResourceName)
+
+					// Call the logger.
+					if l != nil {
+						l.logger.Log(&pblog)
+					} else {
+						log.Infof("Envoy: Orphan Access log message for %s: %s", pblog.CiliumResourceName, pblog.String())
+					}
+				}
+				log.Info("Envoy: Access log closing")
+				conn.Close()
+			}(uc)
+		}
+	}(e.accessLogListener)
```

Comment:
Out of curiosity, why did you pass this listener to the goroutine but not `e.lds`?

Reply:
I guess I have mixed styles here... maybe I thought the "e.accessLogListener" is long or something :-). I'll remove the explicit parameter passing.